File: quests/chapter10/go/security.go
// Chapter 10: Citadel of Firewalls - Security
// Learn about password hashing, JWT validation, and XSS prevention

package main

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"strings"
)

// HashPassword creates a simple hash of password + salt using SHA256
// Note: This is for educational purposes only - use bcrypt in production!
// Example: HashPassword("secret", "randomsalt") -> hex string
func HashPassword(password, salt string) string {
	// TODO: Combine password and salt
	// TODO: Create SHA256 hash of the combined string
	// TODO: Return hex-encoded hash string
	// Hint: Use crypto/sha256 and encoding/hex packages
	return ""
}

// VerifyPassword checks if a password matches a hash
// Returns true if hash(password + salt) equals the provided hash
func VerifyPassword(password, salt, hash string) bool {
	// TODO: Hash the password with salt
	// TODO: Compare with provided hash
	// TODO: Return true if they match
	return false
}

// SanitizeInput removes/escapes dangerous characters to prevent XSS attacks
// Should escape: < > " ' & and remove script tags
// Example: "<script>alert('xss')</script>" -> ""
// Example: "<b>Hello</b>" -> "Hello"
func SanitizeInput(input string) string {
	// TODO: Remove script tags and their content
	// TODO: Remove other HTML tags
	// TODO: Escape special characters: " -> &quot;, ' -> &#39;
	// Hint: Use strings.Replace or regexp for tag removal
	return input
}

// ValidateJWTStructure checks if a JWT token has valid structure
// JWT format: header.payload.signature (3 base64-encoded parts separated by dots)
// This only validates structure, not the signature!
func ValidateJWTStructure(token string) bool {
	// TODO: Split token by dots
	// TODO: Check that there are exactly 3 parts
	// TODO: Verify each part is non-empty
	// TODO: Optionally verify base64 encoding is valid
	return false
}

// DecodeJWTPayload extracts and decodes the payload from a JWT token
// JWT format: header.payload.signature
// Returns the decoded payload as a map, or error if invalid
// Note: This does NOT verify the signature - for educational purposes only!
func DecodeJWTPayload(token string) (map[string]interface{}, error) {
	// TODO: Validate JWT structure first
	// TODO: Split token and get the payload (second part)
	// TODO: Base64 decode the payload (use URL-safe base64)
	// TODO: JSON unmarshal into map[string]interface{}
	// TODO: Return the payload map or error
	return nil, nil
}

// Helper variables for reference (used in solutions)
var (
	_ = sha256.New
	_ = hex.EncodeToString
	_ = base64.URLEncoding
	_ = json.Unmarshal
	_ = strings.Split
)

File: quests/chapter10/rust/security.rs
// Chapter 10: Citadel of Firewalls - Security
// Learn about password hashing, JWT validation, and XSS prevention

use std::collections::HashMap;

/// Hash a password with a salt using SHA256
/// Note: This is for educational purposes only - use bcrypt/argon2 in production!
/// Example: hash_password("secret", "randomsalt") -> hex string
pub fn hash_password(password: &str, salt: &str) -> String {
    // TODO: Combine password and salt
    // TODO: Create SHA256 hash of the combined string
    // TODO: Return hex-encoded hash string
    // Hint: Implement simple SHA256 or use a basic hash function
    String::new()
}

/// Verify if a password matches a hash
/// Returns true if hash(password + salt) equals the provided hash
pub fn verify_password(password: &str, salt: &str, hash: &str) -> bool {
    // TODO: Hash the password with salt
    // TODO: Compare with provided hash
    // TODO: Return true if they match
    false
}

/// Sanitize input to prevent XSS attacks
/// Should escape: < > " ' and remove script tags
/// Example: "<script>alert('xss')</script>" -> ""
/// Example: "<b>Hello</b>" -> "Hello"
pub fn sanitize_input(input: &str) -> String {
    // TODO: Remove script tags and their content
    // TODO: Remove other HTML tags
    // TODO: Escape special characters: " -> &quot;, ' -> &#39;
    // Hint: Use string replacement or regex for tag removal
    input.to_string()
}

/// Check if a JWT token has valid structure
/// JWT format: header.payload.signature (3 base64-encoded parts separated by dots)
/// This only validates structure, not the signature!
pub fn validate_jwt_structure(token: &str) -> bool {
    // TODO: Split token by dots
    // TODO: Check that there are exactly 3 parts
    // TODO: Verify each part is non-empty
    // TODO: Optionally verify base64 encoding is valid
    false
}

/// Extract and decode the payload from a JWT token
/// JWT format: header.payload.signature
/// Returns the decoded payload as a HashMap<String, String>, or error if invalid
/// Note: This does NOT verify the signature - for educational purposes only!
/// Note: For simplicity, this returns string values only (no nested JSON)
pub fn decode_jwt_payload(token: &str) -> Result<HashMap<String, String>, String> {
    // TODO: Validate JWT structure first
    // TODO: Split token and get the payload (second part)
    // TODO: Base64 decode the payload (use URL-safe base64)
    // TODO: Parse JSON into HashMap (simple string values)
    // TODO: Return the payload map or error
    Err("Not implemented".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hash_password_not_empty() {
        let hash = hash_password("password", "salt");
        assert!(!hash.is_empty(), "Hash should not be empty");
    }

    #[test]
    fn test_hash_password_length() {
        let hash = hash_password("password", "salt");
        assert_eq!(hash.len(), 64, "SHA256 hex should be 64 chars");
    }

    #[test]
    fn test_hash_consistency() {
        let hash1 = hash_password("test", "salt");
        let hash2 = hash_password("test", "salt");
        assert_eq!(hash1, hash2, "Same inputs should produce same hash");
    }

    #[test]
    fn test_verify_password_correct() {
        let hash = hash_password("password123", "salt");
        assert!(verify_password("password123", "salt", &hash));
    }

    #[test]
    fn test_verify_password_wrong() {
        let hash = hash_password("password123", "salt");
        assert!(!verify_password("wrongpassword", "salt", &hash));
    }

    #[test]
    fn test_sanitize_removes_script() {
        let result = sanitize_input("<script>alert('xss')</script>");
        assert!(!result.contains("<script>"));
        assert!(!result.contains("alert"));
    }

    #[test]
    fn test_sanitize_removes_html() {
        let result = sanitize_input("<b>Bold</b>");
        assert!(result.contains("Bold"));
        assert!(!result.contains("<b>"));
    }

    #[test]
    fn test_sanitize_escapes_quotes() {
        let result = sanitize_input("Say \"Hello\"");
        assert!(result.contains("&quot;"));
    }

    #[test]
    fn test_jwt_structure_valid() {
        assert!(validate_jwt_structure("header.payload.signature"));
    }

    #[test]
    fn test_jwt_structure_invalid() {
        assert!(!validate_jwt_structure("only.two"));
        assert!(!validate_jwt_structure("one"));
        assert!(!validate_jwt_structure(""));
    }
}
